cirq/ops/three_qubit_gates.py

```diff
@@ -145,6 +144,125 @@ def __str__(self) -> str:
+        self._diag_angles.append(a)
+        self._diag_angles.append(b)
+        self._diag_angles.append(c)
+        self._diag_angles.append(d)
+        self._diag_angles.append(e)
+        self._diag_angles.append(f)
+        self._diag_angles.append(g)
+        self._diag_angles.append(h)
+        self._matrix = np.diag(
+            [np.exp(1j * angle) for angle in self._diag_angles])
+
+    def _unitary_(self) -> np.ndarray:
+        return self._matrix
+
+    def _circuit_diagram_info_(self, args: protocols.CircuitDiagramInfoArgs
+                              ) -> protocols.CircuitDiagramInfo:
+        return protocols.CircuitDiagramInfo(('diag', '#2', '#3'))
+
+    def __pow__(self, exponent: Any) -> 'ThreeQubitDiagonalGate':
+        if not isinstance(exponent, (int, float)):
+            return NotImplemented
+        return ThreeQubitDiagonalGate(
+            *[angle * exponent for angle in self._diag_angles])
+
+    def _decompose_(self, qubits):
+        """An adjacency-respecting decomposition.
+
+        0: ───p_0───@──────────────@───────@──────────@──────────
+                    │              │       │          │
+        1: ───p_1───X───@───p_3────X───@───X──────@───X──────@───
+                        │              │          │          │
+        2: ───p_2───────X───p_4────────X───p_5────X───p_6────X───
+
+        where p_i = T**(4*x_i) and x_i solve the system of equations
+                    [0, 0, 1, 0, 1, 1, 1][x_0]   [l_1]
+                    [0, 1, 0, 1, 1, 0, 1][x_1]   [l_2]
+                    [0, 1, 1, 1, 0, 1, 0][x_2]   [l_3]
+                    [1, 0, 0, 1, 1, 1, 0][x_3] = [l_4]
+                    [1, 0, 1, 1, 0, 0, 1][x_4]   [l_5]
+                    [1, 1, 0, 0, 0, 1, 1][x_5]   [l_6]
+                    [1, 1, 1, 0, 1, 0, 0][x_6]   [l_7]
+        where l_i is self._diag_angles[i].
+
+        The above system was created by equating the composition of the gates
+        in the circuit diagram to np.diag(self._diag_angles) (shifted by a
+        global phase of np.exp(-1j * self._diag_angles[0])).
+        """
+
+        a, b, c = qubits
+        if hasattr(b, 'is_adjacent'):
+            if not b.is_adjacent(a):
+                b, c = c, b
+            elif not b.is_adjacent(c):
+                a, b = b, a
+        sweep_abc = [common_gates.CNOT(a, b), common_gates.CNOT(b, c)]
+        phase_matrix = np.array([[0, 0, 1, 0, 1, 1, 1], [0, 1, 0, 1, 1, 0, 1],
+                                 [0, 1, 1, 1, 0, 1, 0], [1, 0, 0, 1, 1, 1, 0],
+                                 [1, 0, 1, 1, 0, 0, 1], [1, 1, 0, 0, 0, 1, 1],
+                                 [1, 1, 1, 0, 1, 0, 0]])
+        shifted_angles_tail = [
+            angle - self._diag_angles[0] for angle in self._diag_angles[1:]
+        ]
+        phase_solutions = np.linalg.solve(phase_matrix, shifted_angles_tail)
+        p_gates = []
+        for solution in phase_solutions:
+            p_gates.append(common_gates.T**(4 * solution / np.pi))
```

Comment:
Use Z instead of T to drop the factor of four.

Reply:
Done